File: app/src/main/java/bai/bai/bai/demo/view/KeyBoardLinearLayout.java
package bai.bai.bai.demo.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.BaseAdapter;
import android.widget.GridView;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Random;

import bai.bai.bai.demo.R;

public class KeyBoardLinearLayout extends LinearLayout implements AdapterView.OnItemClickListener, View.OnClickListener {

    private int mItemHeight;//行高
    private float mItemTextSize;//item文字大小
    private int mKeyBoardType;//键盘类型
    private KeyBoardNumAdapter mAdapter;

    private static final int TYPE_CALCULATION = 1;//主界面键盘
    private static final int TYPE_PASSWORD = 2;//输入密码
    private static final int TYPE_OTHER = 3;//乱序

    private KeyBoartTouchListener mKeyBoartTouchListener;


    public KeyBoardLinearLayout(Context context) {
        this(context, null);
    }

    public KeyBoardLinearLayout(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        View view = LayoutInflater.from(context).inflate(R.layout.keyboard_linearlayout, null);
        setLinearLayoutStyle(context, attrs, view);
        addView(view);

    }

    private void setLinearLayoutStyle(Context context, AttributeSet attrs, View view) {
        TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.KeyBoardLinearLayout);
        mItemHeight = array.getDimensionPixelSize(R.styleable.KeyBoardLinearLayout_itemHeight, 100);
        mItemTextSize = array.getDimension(R.styleable.KeyBoardLinearLayout_itemTextSize, 30);
        mKeyBoardType = array.getInt(R.styleable.KeyBoardLinearLayout_keyBoardType, TYPE_CALCULATION);

        GridView gvKeyBoardNum = (GridView) view.findViewById(R.id.gr_keyboard_num);
        gvKeyBoardNum.setOnItemClickListener(this);
        mAdapter = new KeyBoardNumAdapter(context, getNumbers(mKeyBoardType));
        gvKeyBoardNum.setAdapter(mAdapter);

        LinearLayout llKeyBoardEmpty = (LinearLayout) view.findViewById(R.id.ll_keyboard_del);
        llKeyBoardEmpty.setOnClickListener(this);

        TextView tvKeyBoardOk = (TextView) view.findViewById(R.id.tv_keyboard_ok);
        tvKeyBoardOk.setOnClickListener(this);
        tvKeyBoardOk.setTextSize(mItemTextSize + 2);
    }

    /**
     * 获得键盘上要填入的数据
     */
    private List<String> getNumbers(int keyBoardType) {
        List<String> numbers = new ArrayList<>();
        switch (keyBoardType) {
            case TYPE_CALCULATION:
                numbers = Arrays.asList(new String[]{"1", "2", "3", "4", "5", "6", "7", "8", "9", ".", "0", "C"});
                break;
            case TYPE_PASSWORD:
                numbers = Arrays.asList(new String[]{"1", "2", "3", "4", "5", "6", "7", "8", "9", "0", " ", "C"});
                break;
            case TYPE_OTHER:
                numbers = Arrays.asList(new String[]{"1", "2", "3", "4", "5", "6", "7", "8", "9", ".", "0", "C"});
                Collections.shuffle(numbers);//乱序
                numbers.set(numbers.indexOf("C"), numbers.get(11));
                numbers.set(11, "C");
                break;
            default:
                numbers = Arrays.asList(new String[]{"1", "2", "3", "4", "5", "6", "7", "8", "9", ".", "0", "C"});
                break;
        }

        return numbers;
    }


    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.ll_keyboard_del:
                mKeyBoartTouchListener.onDelTouch();
                break;
            case R.id.tv_keyboard_ok:
                mKeyBoartTouchListener.onOkTouch();
                break;
        }
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        mKeyBoartTouchListener.onNumberTouch(mAdapter.getItem(position) + "");
    }

    public interface KeyBoartTouchListener {
        void onNumberTouch(String inputNumber);

        void onDelTouch();

        void onOkTouch();
    }

    public void setKeyBoartTouchListener(KeyBoartTouchListener listener) {
        this.mKeyBoartTouchListener = listener;
    }

    private class KeyBoardNumAdapter extends BaseAdapter {

        private Context mContext;
        private List<String> mNumbers;
        private LayoutInflater inflater;

        public KeyBoardNumAdapter(Context mContext, List<String> mNumbers) {
            this.mContext = mContext;
            this.mNumbers = getNumbers(mNumbers);
            this.inflater = LayoutInflater.from(mContext);
        }

        private List<String> getNumbers(List<String> mNumbers) {
            if (mNumbers == null) {
                mNumbers = new ArrayList<>();
            }
            return mNumbers;
        }

        @Override
        public int getCount() {
            return mNumbers.size();
        }

        @Override
        public Object getItem(int position) {
            return mNumbers.get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            ViewHodler hodler;
            if (convertView == null) {
                hodler = new ViewHodler();
                convertView = inflater.inflate(R.layout.item_key_button, null);
                hodler.mTvStyle = (TextView) convertView.findViewById(R.id.tv_style);
                hodler.mTvStyle.setTextSize(mItemTextSize);

                convertView.setTag(hodler);
            } else {
                hodler = (ViewHodler) convertView.getTag();
            }
            setViewHeight(hodler.mTvStyle);

            String number = mNumbers.get(position);
            hodler.mTvStyle.setText(number);

            if (position == 11) {
                hodler.mTvStyle.setTextColor(Color.BLUE);
            }

            return convertView;
        }

        /**
         * 设置子控件的高度
         */
        private void setViewHeight(View view) {
            LayoutParams params = (LayoutParams) view.getLayoutParams();
            params.height = mItemHeight;
            view.setLayoutParams(params);
        }

        private class ViewHodler {
            private TextView mTvStyle;

        }

    }


}

File: app/src/main/java/bai/bai/bai/demo/kt/KeyBoardLinearLayout.kt
package bai.bai.bai.demo.kt

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.BaseAdapter
import android.widget.GridView
import android.widget.LinearLayout
import android.widget.TextView

import java.util.ArrayList
import java.util.Arrays
import java.util.Collections

import bai.bai.bai.demo.R

class KeyBoardLinearLayout constructor(context: Context, attrs: AttributeSet? = null) : LinearLayout(context, attrs), AdapterView.OnItemClickListener, View.OnClickListener {

    private var mItemHeight: Int = 0//行高
    private var mItemTextSize: Float = 0.toFloat()//item文字大小
    private var mKeyBoardType: Int = 0//键盘类型
    private var mAdapter: KeyBoardNumAdapter? = null

    private var mKeyBoartTouchListener: KeyBoartTouchListener? = null

    init {

        val view = LayoutInflater.from(context).inflate(R.layout.keyboard_linearlayout, null)
        setLinearLayoutStyle(context, attrs, view)
        addView(view)

    }

    private fun setLinearLayoutStyle(context: Context, attrs: AttributeSet?, view: View) {
        val array = context.obtainStyledAttributes(attrs, R.styleable.KeyBoardLinearLayout)
        mItemHeight = array.getDimensionPixelSize(R.styleable.KeyBoardLinearLayout_itemHeight, 100)
        mItemTextSize = array.getDimension(R.styleable.KeyBoardLinearLayout_itemTextSize, 30f)
        mKeyBoardType = array.getInt(R.styleable.KeyBoardLinearLayout_keyBoardType, TYPE_CALCULATION)

        val gvKeyBoardNum = view.findViewById<View>(R.id.gr_keyboard_num) as GridView
        gvKeyBoardNum.onItemClickListener = this
        mAdapter = KeyBoardNumAdapter(context, getNumbers(mKeyBoardType))
        gvKeyBoardNum.adapter = mAdapter

        val llKeyBoardEmpty = view.findViewById<View>(R.id.ll_keyboard_del) as LinearLayout
        llKeyBoardEmpty.setOnClickListener(this)

        val tvKeyBoardOk = view.findViewById<View>(R.id.tv_keyboard_ok) as TextView
        tvKeyBoardOk.setOnClickListener(this)
        tvKeyBoardOk.textSize = mItemTextSize + 2
    }

    /**
     * 获得键盘上要填入的数据
     */
    private fun getNumbers(keyBoardType: Int): List<String> {
        var numbers: MutableList<String>? = null
        when (keyBoardType) {
            TYPE_CALCULATION -> numbers = Arrays.asList(*arrayOf("1", "2", "3", "4", "5", "6", "7", "8", "9", ".", "0", "C"))
            TYPE_PASSWORD -> numbers = Arrays.asList(*arrayOf("1", "2", "3", "4", "5", "6", "7", "8", "9", "0", " ", "C"))
            TYPE_OTHER -> {
                numbers = Arrays.asList(*arrayOf("1", "2", "3", "4", "5", "6", "7", "8", "9", " ", "0", "C"))
                Collections.shuffle(numbers)//乱序
                numbers[numbers.indexOf(" ")] = numbers[10]
                numbers[10] = " "
                numbers[numbers.indexOf("C")] = numbers[11]
                numbers[11] = "C"
            }
            else -> numbers = Arrays.asList(*arrayOf("1", "2", "3", "4", "5", "6", "7", "8", "9", "", "0", "C"))
        }

        return numbers
    }


    override fun onClick(v: View) {
        when (v.id) {
            R.id.ll_keyboard_del -> mKeyBoartTouchListener!!.onDelTouch()
            R.id.tv_keyboard_ok -> mKeyBoartTouchListener!!.onOkTouch()
        }
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        mKeyBoartTouchListener!!.onNumberTouch(mAdapter!!.getItem(position).toString() + "")
    }

    interface KeyBoartTouchListener {
        fun onNumberTouch(inputNumber: String)

        fun onDelTouch()

        fun onOkTouch()
    }

    fun setKeyBoartTouchListener(listener: KeyBoartTouchListener) {
        this.mKeyBoartTouchListener = listener
    }

    private inner class KeyBoardNumAdapter(private val mContext: Context, mNumbers: List<String>) : BaseAdapter() {
        private val mNumbers: List<String>
        private val inflater: LayoutInflater

        init {
            this.mNumbers = getNumbers(mNumbers)
            this.inflater = LayoutInflater.from(mContext)
        }

        private fun getNumbers(mNumbers: List<String>?): List<String> {
            var mNumbers = mNumbers
            if (mNumbers == null) {
                mNumbers = ArrayList()
            }
            return mNumbers
        }

        override fun getCount(): Int {
            return mNumbers.size
        }

        override fun getItem(position: Int): Any {
            return mNumbers[position]
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            var convertView = convertView
            val hodler: ViewHodler
            if (convertView == null) {
                hodler = ViewHodler()
                convertView = inflater.inflate(R.layout.item_key_button, null)
                hodler.mTvStyle = convertView!!.findViewById<View>(R.id.tv_style) as TextView
                hodler.mTvStyle!!.textSize = mItemTextSize

                convertView.tag = hodler
            } else {
                hodler = convertView.tag as ViewHodler
            }
            setViewHeight(hodler.mTvStyle!!)

            val number = mNumbers[position]
            hodler.mTvStyle!!.text = number

            if (position == 11) {
                hodler.mTvStyle!!.setTextColor(resources.getColor(R.color.colorAccent))//清除按钮
            }

            return convertView
        }

        /**
         * 设置子控件的高度
         */
        private fun setViewHeight(view: View) {
            val params = view.layoutParams as LinearLayout.LayoutParams
            params.height = mItemHeight
            view.layoutParams = params
        }

        private inner class ViewHodler {
            var mTvStyle: TextView? = null

        }

    }

    companion object {
        private val TYPE_CALCULATION = 1//主界面键盘
        private val TYPE_PASSWORD = 2//输入密码
        private val TYPE_OTHER = 3//乱序
    }


}
